internal/p2p/channel.go

```diff
@@ -0,0 +1,208 @@
+		Error:       errCh,
+	}
+}
+
+func (ch *Channel) Send(ctx context.Context, envelope Envelope) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ch.Out <- envelope:
+		return nil
+	}
+}
+
+func (ch *Channel) SendError(ctx context.Context, pe PeerError) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ch.Error <- pe:
+		return nil
+	}
+}
+
+func (ch *Channel) Receive(ctx context.Context) *ChannelIterator {
+	iter := &ChannelIterator{
+		pipe: make(chan Envelope), // unbuffered
+	}
+	go func() {
+		defer close(iter.pipe)
+		iteratorWorker(ctx, ch, iter.pipe)
+	}()
+	return iter
+}
+
+// ChannelIterator provides a context-aware path for callers
+// (reactors) to process messages from the P2P layer without relying
+// on the implementation details of the P2P layer. Channel provides
+// access to it's Outbound stream as an iterator, and the
+// MergedChannelIterator makes it possible to combine multiple
+// channels into a single iterator.
+type ChannelIterator struct {
+	pipe    chan Envelope
+	current *Envelope
+}
+
+func iteratorWorker(ctx context.Context, ch *Channel, pipe chan Envelope) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case envelope := <-ch.In:
+			select {
+			case <-ctx.Done():
+				return
+			case pipe <- envelope:
+			}
+		}
+	}
+}
+
+// Next returns true when the Envelope value has advanced, and false
+// when the context is canceled or iteration should stop, then Next
+// will return true. In general, use Next, as in:
+//
+//     for iter.Next(ctx) {
+//          envelope := iter.Envelope()
+//          // ... do things ...
+//     }
+//
+func (iter *ChannelIterator) Next(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		iter.current = nil
+		return false
+	case envelope, ok := <-iter.pipe:
+		if !ok {
+			iter.current = nil
+			return false
+		}
+
+		iter.current = &envelope
+
+		return true
+	}
+}
+
+// Envelope returns the current Envelope object held by the
+// iterator. When the last call to Next returned true, Envelope will
+// return a non-nil object. If Next returned false then Envelope is
+// always nil.
+func (iter *ChannelIterator) Envelope() *Envelope { return iter.current }
+
+// MergedChannelIterator produces a Receive iterator from multiple
+// channels to support the reactor use case of needing to consume
+// messages from multiple sources in a single reactor without needing
+// the reactor to manage the concurrency for these handlers.
+func MergedChannelIterator(ctx context.Context, chs ...*Channel) *ChannelIterator {
+	iter := &ChannelIterator{
+		pipe: make(chan Envelope), // unbuffered
+	}
+	wg := &sync.WaitGroup{}
+
+	wait := make(chan struct{})
+	go func() { defer close(wait); wg.Wait() }()
+
+	go func() {
+		defer close(iter.pipe)
+		// we could return early if the context is canceled,
+		// but this is safer because it means the pipe stays
+		// open until all of the ch worker threads end, which
+		// should happen very quickly.
+		<-wait
```

Comment:
Nit: naming suggestion.
```suggestion
	done := make(chan struct{})
	go func() { defer close(done); wg.Wait() }()

	go func() {
		defer close(iter.pipe)
		// we could return early if the context is canceled,
		// but this is safer because it means the pipe stays
		// open until all of the ch worker threads end, which
		// should happen very quickly.
		<-done
```